# Testing a Difference of Means

<!-- JMS: Consider adding test for whether one value matches a known mean. I don't have a great example right now, though -->


One of the most common statistical tests is whether two data sets come from distributions with the same means. The mean is special because of the Central Limit Theorem (see {doc}`Section 8.7.3.2<../08-random-variables/important-continuous-rvs>`) -- regardless of the type of distribution of the underlying data[^clt], the distribution of the sample mean will be approximately Normal if there are at least tens of data points. This allows two approaches to conducting tests involving sample means: we can either use bootstrap resampling or we can use analysis by applying the assumption that sample means are approximately Normal.  To understand how to apply the analytical approaches, we need to characterize the sample mean estimators and the test statistic, which is the difference between the sample mean estimators.



# Characterizing the Difference of Sample Mean Estimators

Suppose we have two  data samples $\mathbf{X}=\left[X_0,X_1,\cdots,X_{n_X-1} \right]$ and $\mathbf{Y}=\left[Y_0,Y_1,\cdots,Y_{n_Y-1}\right]$, where the data samples are assumed to be independent[^ci]. From the results in {doc}`Section 9.5.2<confidence-intervals-estimates>`, the sample mean estimators for these data samples have the following distributions:
\begin{align*}
\hat{\mu}_X &\sim \mbox{Normal} \left(\mu_X, \frac{\sigma_X}{\sqrt{n_X}} \right), \mbox{ and}\\
\hat{\mu}_Y &\sim \mbox{Normal} \left(\mu_Y, \frac{\sigma_Y}{\sqrt{n_Y}} \right). 
\end{align*}
In addition, $\hat{\mu}_X$ and $\hat{\mu}_Y$ are independent. 



[^ci]: Actually, the data within each group are only *conditionally independent* given the group it belongs to.

To assess a difference in means, we will create a test statistic that is the difference between sample mean estimators. For convenience, we will considerr the case $T = \hat{\mu}_X - \hat{\mu}_Y$, but the result for $T = \hat{\mu}_Y - \hat{\mu}_X$ is almost identical. The mean of the test statistic is 
\begin{align*}
E[T] & = E\left[\hat{\mu}_X - \hat{\mu}_Y \right] \\
&= E\left[\hat{\mu}_X\right]  - E \left[\hat{\mu}_Y \right] && \mbox{(by linearity)} \\
&= \mu_X - \mu_Y &&\mbox{(the estimators are unbiased)}.
\end{align*}



To find the variance of $T$, we first rewrite the formula for the test statistic slightly, as $T = \hat{\mu}_X + (-1) \hat{\mu}_Y$. Since $\mathbf{X}$ and $\mathbf{Y}$ are independent, so are variables computed from them. Thus, 
\begin{align*}
\sigma_{T}^2 &= \operatorname{Var}[T] \\
& = \operatorname{Var}\left[\hat{\mu}_X + (-1) \hat{\mu}_Y \right] \\
& = \operatorname{Var}\left[\hat{\mu}_X \right] + 
\operatorname{Var} \left[(-1) \hat{\mu}_Y \right] 
&& \mbox{(by Property 4 of variance)}\\
& = \frac{\sigma_{X}^2}{n_X} +  (-1)^2\frac{\sigma_{Y}^2}{n_Y}
&& \mbox{(from above and Property 3)}\\
& = \frac{\sigma_{X}^2}{n_X} +  \frac{\sigma_{Y}^2}{n_Y} \\
\end{align*}

Importantly, a linear combination of independent Normal random variables is also a Normal random variable. Thus we have a characterization of the test statistic:
\begin{equation*}
T \sim \mbox{Normal} \left( \mu_X - \mu_Y, \sqrt{\frac{\sigma_{X}^2}{n_X} +  \frac{\sigma_{Y}^2}{n_Y}} \right)
\end{equation*}


## Statistical Inference for a Difference of Sample Means *with Known and Equal Variances*

Consider the case where we know that the data come from distributions with the same variance, $\sigma^2$, and that we know the value of $\sigma^2$. (The case of known but unequal variances requires only a minor modification but is rare enough that I omit here.) Suppose we have samples from these distributions, 
\begin{align*}
\mathbf{x} = \left[ x_0, x_1, \ldots, x_{n_X-1} \right],
\end{align*}
and
\begin{align*}
\mathbf{y} = \left[ y_0, y_1,\ldots, y_{n_Y-1} \right].
\end{align*}
Let the averages of the data be denoted by $\overline{\mathbf{x}}$ and $\overline{\mathbf{y}}$, and denote the true (but unknown) means of the distributions as $\mu_X$ and $\mu_Y$.


We can now easily conduct a NHST. The null hypothesis is that the data have the same mean and so $E[T]=0$.  Let the observed value of the test statistic be $t = \overline{\mathbf{x}} - \overline{\mathbf{y}}$. Below I assume that $t>0$; if not, then interchange the role of $x$ and $y$.


Under the assumption that variances of the $X_i$ and $Y_k$ are both equal to $\sigma^2$, the variance of $T$ simplifies to 
\begin{align*}
\sigma_{T}^2  & = \frac{\sigma^2}{n_X} +  \frac{\sigma^2}{n_Y} \\
 & = \sigma^2 \left( \frac{1}{n_X} +  \frac{1}{n_Y} \right) \\
\end{align*}

Once we know the mean and variance of $T$, then the NHST is straight-forward application of the results in {doc}`Chapter 8<../08-random-variables/intro>`:

* For a one-sided hypothesis test, the $p$-value is
    
    $$
    P(T \geq t | H_0) = Q\left(\frac{t }{\sigma_T}\right) = Q\left(\frac{t}{\sigma \sqrt{\frac{1}{n_X}+\frac{1}{n_Y}}}\right).
    $$
    
* For a two-sided NHST, the $p$ value is
    
    $$
    P(|T| \geq t | H_0) = 2 Q\left(\frac{t}{\sigma \sqrt{\frac{1}{n_X}+\frac{1}{n_Y}}}\right).
    $$
Note that in both these cases, the only information that is needed is the observed value of the test statistic, the standard deviation of the data, and the number of samples in each group ($n_X$ and $n_Y$).



## Binary Hypothesis Tests with *Unknown Variance*

In most cases, the variance(s) of the underlying distributions are not known and must be estimated from the data. As in the previous section, scaling the test statistic based on the sample variance will result in a statistic that is no longer Normal. The the scaled test statistic depends on what we can assume about the relationship of the variances of the two samples. If the samples can be assumed to have the same variance, then the test statistic can be scaled so that the resulting variable has a Student's $t$ distribution. If the variances of the data samples are unequal, then the test statistic can be scaled so that the resulting distribution is well approximated by a Student's $t$ distribution.  Before considering each of these cases, we start with some preliminaries:

We wish to determine the probability of observing a difference in averages with magnitude as large as $t = \overline{x} -\overline{y}$ under the null hypothesis, under which the mean of the scaled mean estimator is zero.  To calculate this probability, we will need the variance estimates for each sample, which in turn depend on the mean estimators . The mean estimators are given by 
\begin{align*}
\hat{\mu}_X &= \frac{1}{n_x} \sum_{i=0}^{n_x-1} X_i, \mbox{ and} \\
\hat{\mu}_Y &= \frac{1}{n_y} \sum_{i=0}^{n_y-1} Y_i.
\end{align*}

The unbiased variance estimator for each sample is
\begin{align*}
S_{x}^{2} &= \frac{1}{n_x-1} \sum_{i=0}^{n_x-1} \left( X_{i} - \hat{\mu}_X
\right)^2, \mbox{ and} \\
S_{y}^{2} &= \frac{1}{n_y-1} \sum_{i=0}^{n_y-1} \left( Y_{i} - \hat{\mu}_Y
\right)^2.
\end{align*}


Now we consider the two cases: 

**1.** Test for equal means with *unknown, equal variances*

If the variances can assumed to be equal, then we only need a single estimate of the variance, called the pooled variance estimator. Since we do not know whether the means are equal, then the pooled variance estimator must use the mean estimator for each data sample when centralizing the data for that sample. Because two mean estimators are used, the degrees of freedom adjustment requires us to divide by $n_x + n_y-2$ when calculating the unbiased sample variance estimator for the pooled data. The resulting pooled variance can be written in either of the following forms:
\begin{align*}
S_{p}^2 &= \frac{1}{n_x +n_y -2} \left[  \sum_{i=0}^{n_x-1} \left( X_{i} - \hat{\mu}_X
\right)^2 
+ \sum_{i=0}^{n_y-1} \left( Y_{i} - \hat{\mu}_Y
\right)^2 \right] \\
&= \frac{(n_x-1) S_{X}^2 + (n_y-1) S_{Y}^2}{n_x +n_y -2}
\end{align*}

Then 
\begin{equation*}
\frac{T}{S_p \sqrt { \frac 1 {n_x} + \frac 1 {n_Y}}} \sim t_\nu
\end{equation*}
for $\nu = n_x+n_y-2$ degrees of freedom. 




**2.** Test for equal means with *unknown, unequal varaiances*

If we cannot assume that the variances are equal, then distribution of the following normalized form will be approximately equal to a Student's $t$ distribution:

\begin{align*}
\frac{T}{S_d} \sim t_\nu,
\end{align*}
where 
```{math}
:label: t_denom_unequal
S_d= \sqrt{\frac {S_{x}^2}{ n_x} + \frac {S_{y}^2}{n_y}}.
```
Here, the value of $\nu$ must be determined from the sample standard errors of the means, which
we denote by
\begin{align*}
s_{\overline{x}}^2 &=  {s_{x}^2}/{n_x}, \mbox{ and} \\
s_{\overline{y}}^2 &=  {s_{y}^2}/{n_x}, \mbox{ and} \\
\end{align*}

Then $\nu$ is the largest integer that satisfies
\begin{equation*}
\nu \le \frac{ \left( s_{\overline{x}}^2   +  s_{\overline{y}}^2  \right)^2 }
{ {s_{\overline{x}}^2  }/{(n_x-1)} + {  s_{\overline{y}}^2  }/{(n_y-1)}
}.
\end{equation*}


**Example: Analytical Test on Difference of Means (T-Test)**

```{warning}

The following example is about deaths caused by firearms, and this may be a sensitive topic for some readers. In addition, the example is particularly about assessing the effect of gun legislation on firearms deaths, which is a politically sensitive topic. However, I have chosen to include this example because of its relevance to the national discussion around these topics. Readers should know that a simple analysis like that included here can suggest a relationship between factors that may actually be attributable to other underlying causes. 

Please read this example with an open mind and the desire to see what the data indicates. 
```

In this example, we will consider the effect of "permitless carry" on different types of firearms deaths. Here "permitless carry" (variants of which are also known as "constitiutional carry") allows gun owners in a state to carry a loaded firearm without having to apply to the government for a gun permit. There are actually many variations on permitless carry, but I consider a state to allow permitless carry if state's citizens can generaly carry (either open or concealed) a loaded handgun without a gun permit. 

The two research hypotheses I will consider are:
1. There is a difference in firearms homicide rates between states with permitless carry and those without.
1. There is a difference in firearms suicide rates between states with permitless carry and those without.

To answer these questions, we will need to collect data from two separate sources. For the firearms mortality data, we can use firearms mortality data from CDC WONDER, which is the Wide-ranging ONline Data for Epidemiologic Research.  In particular, I have used WONDER SEARCH to access data from [Underlying Cause of Death 1999-2020](https://wonder.cdc.gov/ucd-icd10.html). I have downloaded data by state for the following Causes of Death, shown using IC-10 codes:
* X93: Assault by handgun discharge
* X94: Assault by rifle, shotgun, and larger firearm discharge
* X95: Assault by other and unspecified firearm discharge.

I selected to download the total deaths across these categories, the population, and the crude rate. Here, the crude rate is defined as the deaths per population times 100,000. The resulting download is a tab-separate value (TSV) file, which is very similar to the CSV files that we saw previously, except the field separators are tabs instead of commas. The resulting file is available on GitHub at https://raw.githubusercontent.com/jmshea/Foundations-of-Data-Science-with-Python/main/09-moments/data/wonder-homicides-2020.tsv

Data for suicides was retrieved from the same WONDER database using the following IC-10 codes:
* X72: Intentional self-harm by handgun discharge
* X73: Intentional self-harm by rifle, shotgun, and larger firearm discharge
* X74: Intentional self-harm by  other and unspecified firearm discharge.
The resulting TSV is available on GitHub at https://raw.githubusercontent.com/jmshea/Foundations-of-Data-Science-with-Python/main/09-moments/data/wonder-suicides-2020.tsv.

To load these data into Pandas, we can use `pd.read_csv()`, but we need to tell that function that the data is separated by tabs instead of commas. To do this, we will pass the keyword argument `sep = '\t'`, where `\t` is a special code that translates to the tab character. Code to load these two data sets is below. (Note that because these files are being read directly from GitHub, you will need an active internet connection with GitHub access to retrieve them.)

First, let's load the libraries we will need:

In [27]:
import pandas as pd
import numpy as np
import numpy.random as npr
import scipy.stats as stats

In [3]:
#suicides = pd.read_csv('data/wonder-suicides-2020.tsv', sep='\t')
#homicides = pd.read_csv('data/wonder-homicides-2020.tsv', sep='\t')
fdsp_prefix = 'https://raw.githubusercontent.com/jmshea/Foundations-of-Data-Science-with-Python/'
suicides = pd.read_csv(fdsp_prefix + 'main/09-moments/data/wonder-suicides-2020.tsv', sep='\t')
homicides= pd.read_csv(fdsp_prefix + 'main/09-moments/data/wonder-homicides-2020.tsv', sep='\t')

Let's merge these two dataframes into a single dataframe. To do that, I am first going to do some preparation:
1. We will drop the Notes, Crude Rate, and State Code columns from the `suicides` dataframe.
2. We will drop all of the above plus the Population column from the `homicides` dataframe (because we are going to merge with the `suicides` dataframe that already has the information.
3. We will relabel the `Deaths` column of each dataframe to match the type of death.
Here are these first steps:

In [4]:
#1.
suicides.drop(columns=['Notes', 'Crude Rate', 'State Code'], inplace=True)

#2.
homicides.drop(columns=['Notes', 'Population', 'Crude Rate', 'State Code'], inplace=True)

#3. 
suicides.rename({'Deaths': 'Suicides'}, axis=1, inplace=True)
homicides.rename({'Deaths': 'Homicides'}, axis=1, inplace=True)

Let's check each dataframe now:

In [5]:
suicides.head()

,State,Suicides,Population
0,Alabama,542,4921532
1,Alaska,133,731158
2,Arizona,830,7421401
3,Arkansas,364,3030522
4,California,1552,39368078


In [6]:
homicides.head()

,State,Homicides
0,Alabama,564
1,Alaska,27
2,Arizona,382
3,Arkansas,282
4,California,1731


A few notes:
1. We are interested in mortality **rates**, and we could have preserved the `Crude Rate` column instead of the `Deaths` column. However, the `Crude Rate` is computed from `Deaths` and `Population`, and preserving these separately will allow us to analyze the data in different ways. Moreover, the `Crude Rate` for suicide in some states is listed as `unreliable`, corresponding to fewer than 20 suicide deaths in that state. Since we are never using these rates for a single state in isolation, the values are useful to our analysis, and we will use all the rate data.
2. Note that these dataframes are of different sizes:

In [7]:
len(suicides), len(homicides)

(50, 49)

When we merge them, we need to decide to handle any discrepancies in which states are included. I will use the approach that we only include in the combined data set those states that have entries in *both* of the `suicides` and `homicides` dataframes. This is called an *inner join*.  We will use the `merge()` method of the `suicides` Pandas dataframe, which takes as argument the dataframe to be merged. We will specify the keyword argument `on = 'State'` to specify that we are matching up the rows from the different dataframes based on the entry in the `State` column. We will  pass the keyword argument `how = 'inner'`, to do an inner join and preserve only those entries that appear in *both* of the dataframes being merged.

In [8]:
all_deaths=suicides.merge(homicides, on='State', how='inner')
all_deaths.head()

,State,Suicides,Population,Homicides
0,Alabama,542,4921532,564
1,Alaska,133,731158,27
2,Arizona,830,7421401,382
3,Arkansas,364,3030522,282
4,California,1552,39368078,1731


Let's check the length of the merged dataframe:

In [9]:
len(all_deaths)

48

Even though the smaller of the dataframes had 49 rows, the inner join produced only 48 rows. The `homicides` dataframe actually has entries for all 50 states, but the `suicides` dataframe has entries for only 48 states and the District of Columbia.

Now let's go ahead and compute the homicide and suicide rates (scaled up by 100,000):

In [10]:
all_deaths['Homicide Rate'] = all_deaths['Homicides'] / all_deaths['Population'] * 100_000

In [11]:
all_deaths['Suicide Rate'] = all_deaths['Suicides'] / all_deaths['Population'] * 100_000

In [12]:
all_deaths.head()

,State,Suicides,Population,Homicides,Homicide Rate,Suicide Rate
0,Alabama,542,4921532,564,11.459846,11.012831
1,Alaska,133,731158,27,3.692772,18.190323
2,Arizona,830,7421401,382,5.147276,11.183872
3,Arkansas,364,3030522,282,9.305328,12.011132
4,California,1552,39368078,1731,4.396963,3.942280


The second source of data we will need is on permitless carry of firearms. I used the table at [Wikipedia: Constitutional Carry - Ages to carry without a permit](https://en.wikipedia.org/wiki/Constitutional_carry#Ages_to_carry_without_a_permit) and the source documents to create a CSV file that gives a list states that permitted a *state resident* to carry a *handgun* without a permit as of 2020. For each such state, the age at which a handgun can be carried without a permit is listed for both *open* and *concealed* carry. If permitless carry of a handgun by a state resident is not allowed for one of these categories, the entry is "N/A". States not in this CSV file do not allow permitless carry as of 2020.  The resulting CSV file is available on GitHub at https://raw.githubusercontent.com/jmshea/Foundations-of-Data-Science-with-Python/main/09-moments/data/permitless-carry-2020.csv.

Let's load this data into another Pandas dataframe:

In [13]:
permitless_df = pd.read_csv(fdsp_prefix + 'main/09-moments/data/permitless-carry-2020.csv')
permitless_df

,State,Permitless_open,Permitless_concealed
0,Alabama,18.0,NaN
1,Alaska,16.0,21.0
2,Arizona,18.0,21.0
3,Arkansas,18.0,18.0
4,Idaho,18.0,18.0
5,Kansas,18.0,21.0
6,Kentucky,18.0,21.0
7,Maine,18.0,21.0
8,Mississippi,18.0,18.0
9,Missouri,18.0,18.0


We will again use the `merge()` method of the Pandas dataframe, 
but this time we will use it on the `all_deaths` dataframe.  We will specify the keyword argument `on = 'State'` to specify that we are matching up the rows from the different dataframes based on the entry in the `State` column. For this merge operation, we do not want to do an inner join because that would drop all of the death data for states that do not allow permitless carry. Instead, we will perform a *left join*, which means we will preserve all of the keys in the `all_deaths` dataframe, which appears to the left of the `permitless` dataframe.

Here is that left join:

In [14]:
df = all_deaths.merge(permitless_df, on = 'State', how = 'left')

It will be convenient to index this dataframe by the State:

In [15]:
df.set_index('State', inplace=True)

For convenience of display, I will remap the order of the columns:

In [16]:
df2 = df[ ['Population', 'Homicides', 'Homicide Rate', 'Suicides', 'Suicide Rate', 'Permitless_open', 'Permitless_concealed'] ]

In [17]:
df2.head()

,Population,Homicides,Homicide Rate,Suicides,Suicide Rate,Permitless_open,Permitless_concealed
State,,,,,,,
Alabama,4921532,564,11.459846,542,11.012831,18.0,NaN
Alaska,731158,27,3.692772,133,18.190323,16.0,21.0
Arizona,7421401,382,5.147276,830,11.183872,18.0,21.0
Arkansas,3030522,282,9.305328,364,12.011132,18.0,18.0
California,39368078,1731,4.396963,1552,3.942280,NaN,NaN


Note that California is not one of the states in the `permitless` dataframe and yet its firearms mortality data is preserved in the merged dataframe.

Finally, let's split `df` back into two separate dataframes based on whether they allow permitless carry (open or concealed, at any age) or not. When using `df.query()`, we can combine logical conditions using "|" to represent logical **or** or "&" to represent logical **and**. Thus, the following queries can be used to partition `df`:

In [18]:
permitless = df2.query('Permitless_concealed >0 | Permitless_open >0')
permit = df2.query('Permitless_concealed.isnull() & Permitless_open.isnull()')

As a check, we can see that the sizes of `permitless` and `permit` equal the size of `df`:

In [19]:
len(permitless), len(permit), len(df)

(17, 31, 48)

**Effect of Permitless Carry on Homicide Rates***

Let's conduct a test to see whether the average homicide rate differs between permitless carry states and states that require a permit to carry a gun. In this analysis, we will use the simplest approach, which is to directly compute the average of the homicide rates. We discuss an alternative approach in the problems further below.

Computing the average homicide rate for each class of states is easy:

In [20]:
permitless['Homicide Rate'].mean()

6.066280811037098

In [21]:
permit['Homicide Rate'].mean()

5.391510100924644

The difference in means is

In [59]:
Hdiff = permitless['Homicide Rate'].mean() - permit['Homicide Rate'].mean()
Hdiff

0.6747707101124538

We can see that the homicide rate for permitless carry states is higher. We can perform bootstrap resampling to determine if the observed difference is statistically significant. Since our initial research hypothesis is that there is a difference in homicide rates between permitless carry states and those without, it makes sense to carry out a two-sided test. Our null hypothesis is that there is no difference in homicide rate between these two classes of states, and so we will pool the homicide rate data for all states and draw random bootstrap samples representing each class of states. We then determine the relative frequency of observing a difference in averages as high as the one present in the data. Because we are interested in conducting statistical tests on both homicide rates and suicide rates and most of the programming code is identical, we write a function that can carry out a bootstrapping test on any column of the data:

In [79]:
def bootstrap_permit (df, permitless, permit, column_name, num_sims = 10_000):
  pooled=df[column_name]
  permitless_len = len(permitless)
  permit_len = len(permit)

  diff = permitless[column_name].mean() - permit[column_name].mean()
  print(f'Observed difference in means was {diff:.2f}')
  count = 0

  for sim in range(num_sims):
    # Draw the bootstrap samples
    bs_permitless = npr.choice(pooled, permitless_len)
    bs_permit =npr.choice(pooled, permit_len)

    # Now compute the statistic for the bootstrap samples
    bs_t =  bs_permitless.mean() - bs_permit.mean() 

    # And conduct a two-sided test
    if abs(bs_t) >= diff:
      count+=1

  print(f'Prob. of observing absolute difference as large as data =~ {count/num_sims: .2g}')

Then we can carry out the two-sided hypothesis test for a difference of means as follows:

In [80]:
bootstrap_permit(df2, permitless, permit, 'Homicide Rate')

Observed difference in means was 0.67
Prob. of observing absolute difference as large as data =~  0.54


Thus, under the null hypothesis that the means are identical, we see a difference in homicide rates this large more than 50% of the time. The observed difference in means is not statistically significant.

As an alternative to conducting the NHST using bootstrap resampling, we can conduct an analytical $T$-test. We cannot assume that the data has the same variance, so we calculate the scaling factor to make the sample mean estimator into a Student's $t$ random variable using {eq}`t_denom_unequal`:

In [113]:
sem_x = stats.sem(permitless['Homicide Rate'])
sem_y = stats.sem(permit['Homicide Rate'])
sem_x, sem_y

(1.0171744597615262, 0.5861643161996805)

In [114]:
nx = len (permitless)
ny = len(permit)
nx, ny

(17, 31)

In [115]:
Sd = np.sqrt(sem_x**2 + sem_y**2)
Sd

1.1739814679870342

In [116]:
nu = (sem_x**2 +sem_y**2) ** 2 / ( sem_x**4/(nx-1) + sem_y**4/(ny-1) )
nu

26.81406900531913

In [117]:
HT = stats.t(df = nu, scale = Sd)

In [118]:
2*HT.sf(Hdiff)

0.5702375898678336

The analytical result gives a very similar value to the bootstrap result. This test for a difference of means is often just called a T-test, and it is common enough that SciPy.stats has a built-in method for carrying out the analytical T-test, `stats.ttest_ind()`. Here, the "ind" refers to the fact that the samples must be *independent*. The call signature for this function is shown below:

    scipy.stats.ttest_ind(a, b, axis=0, equal_var=True, nan_policy='propagate', permutations=None, random_state=None, alternative='two-sided', trim=0)
    
Here `a` and `b` are the two data samples. The main thing that we will need to be aware of here is that the default is to assume that the variances are **equal**, which we cannot assume for our data set; so we will need to pass the keyword argument `equal_var=False`:

In [119]:
stats.ttest_ind(permitless['Homicide Rate'], permit['Homicide Rate'], equal_var=False)

Ttest_indResult(statistic=0.5747711769841217, pvalue=0.570237589867833)

The $p$-value matches our analytical result exactly because it is using the exact same formulas internally.

**Effect of Permitless Carry on Suicide Rates**

Now let's consider whether permitless carry affects suicide rates. As before, we start by computing the mean suicide rates for both the permitless and the permit group, as well as the difference in means:

In [72]:
permitless['Suicide Rate'].mean()

12.555980404372416

In [73]:
permit['Suicide Rate'].mean()

7.238123660784346

The difference in means is

In [74]:
Sdiff = permitless['Suicide Rate'].mean() - permit['Suicide Rate'].mean()
Sdiff

5.31785674358807

The difference in mean suicide rates (5.32) is *much* larger than the difference in homicide rates (0.67). The larger difference is much less likely to be attributed to randomness in sampling, but we need to conduct tests to confirm whether the result should be considered statistically significant. We can carry out a two-sided hypothesis test on the difference of means using the function that we created previously:


In [81]:
bootstrap_permit(df2, permitless, permit, 'Suicide Rate')

Observed difference in means was 5.32
Prob. of observing absolute difference as large as data =~  0


In [92]:
bootstrap_permit(df2, permitless, permit, 'Suicide Rate', num_sims=10_000_000)

Observed difference in means was 5.32
Prob. of observing absolute difference as large as data =~  3.5e-05


The probability of observing such a large difference in means under the null hypothesis is so small that most runs of the bootstrap simulation will not generate any events where a difference this large occurs under the null hypothesis, unless we use a much larger number of simulation points.  If we want to accurately estimate the probability of seeing such a large difference under the null hypothesis, the analytical approach can estimate that probability without requiring a huge number bootstrap simulation points. As for the case of homicide rates, we start by finding the sample SEM for each group:

In [83]:
sem_x = stats.sem(permitless['Suicide Rate'])
sem_y = stats.sem(permit['Suicide Rate'])
sem_x, sem_y

(0.8939434974895945, 0.588546275254346)

In [84]:
nx = len (permitless)
ny = len(permit)
nx, ny

(17, 31)

Then the standard error for the decision statistic is

In [85]:
Sd = np.sqrt(sem_x**2 + sem_y**2)
Sd

1.0702904721708462

Next, we need to estimate the number of degrees of freedom to use in the Student's $t$ distribution:

In [89]:
nu = (sem_x**2 +sem_y**2) ** 2 / ( sem_x**4/(nx-1) + sem_y**4/(ny-1) )
nu

29.882264069086958

Putting this all together, we can create the SciPy distribution object as

In [102]:
HT = stats.t(df = nu, scale = Sd)


and the probability of such a large difference under $H_0$ is approximately

In [103]:
2*HT.sf(Sdiff)

2.5702508897424183e-05

For comparison, I ran the bootstrap simulation for 10 million points, and the resulting probability estimate was $3.5 \times 10^{-5}$. Thus, either method produces a similar result, but the analytical solution is much faster in this case. Finally, let's check our analytical result using `stats.ttest_ind()`:

In [104]:
stats.ttest_ind(permitless['Suicide Rate'], permit['Suicide Rate'], equal_var=False)

Ttest_indResult(statistic=4.96861074807289, pvalue=2.5702508897423888e-05)

## Exercises

1. Use the Student's $T$ random variable to determine 95% confidence intervals for the mean difference under the null hypothesis for each of the statistics above. Are the resulting confidence intervals compatible with the observed differences of means?

2. The tests above are based on averaging over states. But the states involved have very different populations. Compute the average rate over populations by summing up the number of homicides for each group and then dividing by the corresponding total population for the group. Conduct a bootstrapping NHST based on the observed rates. Repeat for suicide rates.